[Li Lehao] iP

Li-Lehao opened the pull request:
# Duke x Lehao

> "Your mind is for having ideas, not holding them."
> -- David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to to. It's,
* text-based
* easy to learn
* ~~*Fast*~~ Super Fast to use

All you need to do is,
1. download it from [here](https://github.com/Li-Lehao/ip/releases).
2. double-click it.
3. learn some basic command
4. let it manage your tasks for you 😎

And it is **FREE**

Features:
- [x] Managing tasks, events and deadlines.
- [ ] Get a day's schedule.
- [ ] Search tasks by keyword.

If you Java Programmer, you can use it to practice Java too. Here's the `main` method:
```Java
    public static void main(String[] Args){
        new Duke("data/taskList.txt").run();
    }
```

|  Epoch  |  Iteration  |  Time Elapsed (hh:mm:ss)  |  Mini-batch Accuracy |  Mini-batch Loss  |  Base Learning Rate |
| ---- | ---- | ---- | ---- | ---- | ---- |
|       1 |           1 |       00:00:02 |       29.69% |       2.3961 |          0.0010 |
|       3 |          50 |       00:01:55 |       60.94% |       0.9970 |          0.0010 |
|       5 |         100 |       00:03:59 |       57.81% |       0.8288 |          0.0010 |
|       8 |         150 |       00:06:03 |       68.75% |       0.6458 |          0.0010 |
|      10 |         200 |       00:07:59 |       71.88% |       0.6149 |          0.0010 |
|      12 |         250 |       00:09:51 |       78.12% |       0.5660 |          0.0010 |
|      15 |         300 |       00:11:48 |       78.12% |       0.5468 |          0.0010 |
|      17 |         350 |       00:13:42 |       81.25% |       0.4562 |          0.0010 |
|      20 |         400 |       00:15:42 |       87.50% |       0.3561 |          0.0010 |
|      20 |         420 |       00:16:29 |       89.06% |       0.3140 |          0.0010 |

Commit:
Add Gradle support

Li-Lehao committed:
Level-1. Greet, echo, and exit. Implement a skeletal version of Duke that starts by greeting the user, repeat commands entered by the user, and exits when the user types "bye".

Li-Lehao committed:
Level-2. Add, list. Added the feature to store the text entered by the user and display them back to the user when requested.

Li-Lehao committed:
Level-3. Mark as done. Add the feature to mark the task as done. Added task class and taskNumberOutOfBoundException.

Li-Lehao committed:
Level-4. Added 3 classes: Todo, Deadline, Event. They are subclasses of Task.

Li-Lehao committed:
Level-4. Added 3 classes: Todo, Event, Deadline. They are subclasses of Task.
A-TextUiTesting. Construct an input.txt and an EXPECTED.txt. Pass the automated text UI testing.

Li-Lehao committed:
Level-5. Added handle errors feature. Do regression test.

Li-Lehao committed:
Level-6. Added delete task feature. Using automated text UI testing to test Duke.

Li-Lehao committed:
branch-level-7. Add features: auto load and save task lists from hard disk.

Li-Lehao committed:
branch-level8. Add features: understand date and times.

Li-Lehao committed:
Merge branch 'branch-level-7'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

Li-Lehao committed:
A-MoreOOP.  A-Packages.

Li-Lehao committed:
A-JUnit. Add JUnit tests.

Li-Lehao committed:
A-Jar. Create a JAR file.

Li-Lehao committed:
Level-9. Add features: find tasks by a keyword, find tasks on a certain date

Li-Lehao committed:
A-JavaDoc. A-CodingStandard.

kslui99 commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,58 @@
+
+import java.time.LocalDate;
+import java.time.LocalTime;
+
```
Perhaps you may consider writing a Javadoc header comment describing the class here, since the coding standard mentions that "You MUST write header comments for all classes, public methods except for getters/setters."

I notice that there is no Javadoc comment for your other classes as well, so you may want to consider writing header comments for all the classes.

kslui99 commented on src/main/java/command/ExitCommand.java:
```diff
@@ -0,0 +1,24 @@
+    public boolean isExit() {
+        return true;
+    }
+
```
Maybe you can consider adding a Javadoc comment describing the execute() method here as well? This is to be consistent with your commenting for all the other execute() methods

kslui99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,101 @@
+                }
+            }
+            if (fullCommand.length() >= 6 && fullCommand.substring(0, 4).equals("find")) {
+                String findKeyword = fullCommand.substring(fullCommand.indexOf(" "), fullCommand.length());
```
Should findKeyword be named keywordToFind instead, so that it sounds like a noun instead of a verb?

kslui99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,155 @@
+            sc = new Scanner(f);
+
+            int n = sc.nextInt();
+            //        System.out.println("total number of tasks loaded from file = " + n);
```
Should this line be deleted since it is not used anymore? I notice a few other occurrences of this in this file as well.

kslui99 commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,82 @@
+     * @return A boolean value indicating whether the task happen on that day.
+     */
+    @Override
+    public boolean onDate(LocalDate date) {
```
Maybe this method should be renamed from onDate() to isOnDate(), so that it sounds more like a boolean?

kslui99 reviewed (commented):
Overall, I think you did a great job! I cannot find any other coding standard violations :)

Li-Lehao committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-CodingStandard

Li-Lehao committed:
A-Gradle. A-CheckStyle. Use gradle to build and run Duke. Use gradle to run unit tests. Use checkStyle to detect coding style violations.

Li-Lehao committed:
A-Varargs. Use varargs in ui.printMessages() method. Simplify some printMessages() expressions.

Li-Lehao committed:
GUI.

Li-Lehao committed:
Level-10.

meixuanjin commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,56 @@
+import duke.Ui;
+import task.Task;
+
+public class AddCommand extends Command {
```
Class names could be added to describe what the class is for.

meixuanjin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     * @param time The time of the task.
+     * @return A Task object, the inserted task.
+     */
+    public Task add(Task.TaskType type, String description, LocalDate date, LocalTime time) {
```
Maybe this can be renamed to addTask for more clarity?

meixuanjin reviewed (commented):
Code looks good generally, other than a few long methods with many try/catch blocks. Methods are named well and code is readable!

Li-Lehao committed:
Change build.gradle file to create new JAR

Li-Lehao committed:
Add assertions in Storage.java file

Many problems may occur during file loading and storing, such as file not being found, parent dir does not exist, taskList.txt format is incorrect, and so on.

Adding assertions to this part of the code allow us to better check some exceptions and ensure that the program runs normally.

Li-Lehao committed:
Merge pull request #1 from Li-Lehao/A-Assertions

Add assertions in Storage.java file

Li-Lehao committed:
Improve code quality

There are several methods that redundantly initialize a new Array when calling ui.printMessage which is a vararg method.

To improve code quality as well as efficiency let's remove the redundant array initializing.

Li-Lehao committed:
Merge pull request #2 from Li-Lehao/A-CodeQuality

Improve code quality

Li-Lehao committed:
Use streams in DateCommand.execute() method

DateCommand.execute() method find the tasks on a specific day by scanning through all the tasks in the list. This part of code is a bit too long and the efficiency is low.

Using streams in this method allows us to make the code clear. It also allows us to use parallel stream to speed up this method if possible.

Li-Lehao committed:
Merge pull request #3 from Li-Lehao/A-Streams

Use streams in DateCommand.execute() method

Li-Lehao committed:
Set up Coutinuous Integration  by adding gradle.yml file

Li-Lehao committed:
Add extension C-Sort

Adding a sort-by-time feature allows users to sort their tasks by deadline.

Li-Lehao committed:
Add new feature: sort by time.

This new feature allows us to sort the tasks in the current task list by their time.

Li-Lehao committed:
Merge pull request #4 from Li-Lehao/C-Sort

Add new feature: sort by time.

Li-Lehao committed:
Add FXML

Li-Lehao committed:
test

Li-Lehao committed:
Modify gui

Li-Lehao committed:
A-BetterGui

Li-Lehao committed:
Add user guide and Ui screenshot

Li-Lehao committed:
A-Release

Li-Lehao committed:
Modify the link in README.md

Li-Lehao committed:
Edit gradle.yml file to remove the IO redirection test

since it is not applicable to the current duke project anymore

Li-Lehao committed:
Edit gradle.yml to remove the IO redirection test

since it is not applicable to the current duke project
